src/number_theory/cyclotomic/basic.lean

```diff
@@ -0,0 +1,225 @@
+/-
+Copyright (c) 2021 Riccardo Brasca. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Riccardo Brasca
+-/
+
+import ring_theory.polynomial.cyclotomic.basic
+import number_theory.number_field
+import algebra.char_p.algebra
+
+/-!
+# Cyclotomic extensions
+
+Let `A` and `B` be commutative rings with `algebra A B`. For `S : set ℕ+`, we define a class
+`is_cyclotomic_extension S A B` expressing the fact that `B` is obtained by `A` by adding `n`-th
+primitive roots of unity, for all `n ∈ S`.
+
+## Main definition
+
+* `is_cyclotomic_extension S A B` : means that `B` is obtained by `A` by adding `n`-th primitive
+roots of unity, for all `n ∈ S`.
+
+## Main results
+
+* `is_cyclotomic_extension.trans` : is `is_cyclotomic_extension S A B` and
+`is_cyclotomic_extension T B C`, then `is_cyclotomic_extension (S ∪ T) A C`.
+* `is_cyclotomic_extension.union_right` : given `is_cyclotomic_extension (S ∪ T) A B`, then
+  `is_cyclotomic_extension T (adjoin A { b : B | ∃ a : ℕ+, a ∈ S ∧ b ^ (a : ℕ) = 1 }) B`.
+* `is_cyclotomic_extension.union_right` : given `is_cyclotomic_extension T A B` and `S ⊆ T`, then
+  `is_cyclotomic_extension S A (adjoin A { b : B | ∃ a : ℕ+, a ∈ S ∧ b ^ (a : ℕ) = 1 })`.
+* `is_cyclotomic_extension.finite` : if `S` is finite and `is_cyclotomic_extension S A B`, then
+  `B` is a finite `A`-algebra.
+* `is_cyclotomic_extension.finite` : a finite cyclotomic extension of a number field is a number
+  field.
+
+## Implementation details
+
+Our definition of `is_cyclotomic_extension` is very general, to allow rings of any characteristic
+and infinite extension, but it will mainly be used in the case `S = {n}` with `(n : A) ≠ 0` (and
+for integral domains).
+All results are in the `is_cyclotomic_extension` namespace.
+Note the some lemmas, `is_cyclotomic_extension.trans`, `is_cyclotomic_extension.finite` and
+
+-/
+
+open polynomial algebra finite_dimensional module set
+
+open_locale big_operators
+
+universes u v w z
+
+variables (n : ℕ+) (S T : set ℕ+) (A : Type u) (B : Type v) (K : Type w) (L : Type z)
+variables [comm_ring A] [comm_ring B] [algebra A B]
+variables [field K] [field L] [algebra K L]
+
+noncomputable theory
+
+/-- Given an `A`-algebra `B` and `S : set ℕ+`, we define `is_cyclotomic_extension S A B` requiring
+that `cyclotomic a A` has a root in `B` for all `a ∈ S` and that `B` is generated over `A` by the
+roots of `X ^ n - 1`. -/
+class is_cyclotomic_extension : Prop :=
+(ex_root {a : ℕ+} (ha : a ∈ S) : ∃ r : B, aeval r (cyclotomic a A) = 0)
+(adjoint_roots : ∀ (x : B), x ∈ adjoin A { b : B | ∃ a : ℕ+, a ∈ S ∧ b ^ (a : ℕ) = 1 })
+
+namespace is_cyclotomic_extension
+
+section basic
+
+lemma iff : is_cyclotomic_extension S A B ↔
```

Comment:
This lemma is basically the same as the `is_cyclotomic_extension_iff` generated by adding `@[mk_iff]` to `class is_cyclotomic_extension`, the only difference is for that lemma `a` is implicit. Have you considered just using the autogenerated one instead?

Reply:
Ah yes, this is better.